### ISSUE TITLE
High memory occupation

### ISSUE THREAD

**Issue by hzxGoForward** [sigp/lighthouse#5914]
Yesterday, the memory usage of multiple Lighthouse clients running on the Holesky test network suddenly spiked above 90%, leading to them being forcibly shut down by the system.

Each Lighthouse client is operating on an individual node, with each node configured with 8vCPUs, 32GB of memory, and an upstream bandwidth of 200Mb/s. Below are the graphs of our memory utilization, with the timestamps in Beijing Time (UTC +8 hours), along with the corresponding operation logs, whose timestamps are in UTC. 

The following is memory occupation graph:
![D831B06A-CF7A-420F-97AA-1DF64D86640B](https://github.com/sigp/lighthouse/assets/12388179/985dbbc8-5dd6-4a10-bccf-637b088ba242)

The following is log:
```shell
Jun 10 10:03:06.003 INFO Synced                                  slot: 1842615, block:    …  empty, epoch: 57581, finalized_epoch: 57579, finalized_root: 0x60b1…f872, exec_hash: 0x8cc3…3520 (verified), peers: 99, service: slot_notifier
Jun 10 10:03:10.276 INFO New block received                      root: 0x14d0c19b6bd8c612fc3f29d9444e6d656b6037a385361fa44d7d14b42482bcdc, slot: 1842615
Jun 10 10:03:18.002 INFO Synced                                  slot: 1842616, block:    …  empty, epoch: 57581, finalized_epoch: 57579, finalized_root: 0x60b1…f872, exec_hash: 0x834c…b016 (verified), peers: 106, service: slot_notifier
Jun 10 10:03:30.008 INFO Synced                                  slot: 1842617, block:    …  empty, epoch: 57581, finalized_epoch: 57579, finalized_root: 0x60b1…f872, exec_hash: 0x834c…b016 (verified), peers: 100, service: slot_notifier
Jun 10 10:03:32.881 INFO New block received                      root: 0xbe908fad3cb22e25a8dd01773bc4fb619fdfee8cba0b16973106bd9981fd7129, slot: 1842617
Jun 10 10:03:37.424 INFO New block received                      root: 0x847d7065ecf2e5b2b94d01f2a27ad85adad6199c5b0f8d24961df43a0f432d12, slot: 1842618
Jun 10 10:03:42.009 INFO Synced                                  slot: 1842618, block:    …  empty, epoch: 57581, finalized_epoch: 57579, finalized_root: 0x60b1…f872, exec_hash: 0xffe7…ce05 (verified), peers: 101, service: slot_notifier
Jun 10 10:03:42.207 INFO Gossipsub blob processed, imported fully available block, block_root: 0x847d…2d12
Jun 10 10:03:54.001 INFO Synced                                  slot: 1842619, block:    …  empty, epoch: 57581, finalized_epoch: 57579, finalized_root: 0x60b1…f872, exec_hash: 0xd521…252f (verified), peers: 99, service: slot_notifier
Jun 10 10:04:06.008 INFO Synced                                  slot: 1842620, block:    …  empty, epoch: 57581, finalized_epoch: 57579, finalized_root: 0x60b1…f872, exec_hash: 0xd521…252f (verified), peers: 103, service: slot_notifier
Jun 10 10:04:08.736 INFO New block received                      root: 0xa94d6a79720f0a99199d9e57bcf70eaefbe55b5500c0e63b85653d082fa122a9, slot: 1842620
Jun 10 10:04:13.267 INFO New block received                      root: 0x66e00af7731a3c3184488378e070035b8760d8250c60f382633d3e8a49a0e00c, slot: 1842621
Jun 10 10:04:17.871 INFO Gossipsub blob processed, imported fully available block, block_root: 0x66e0…e00c
Jun 10 10:04:18.003 INFO Synced                                  slot: 1842621, block:    …  empty, epoch: 57581, finalized_epoch: 57579, finalized_root: 0x60b1…f872, exec_hash: 0xcdcc…9922 (verified), peers: 104, service: slot_notifier
Jun 10 10:04:28.256 ERRO Work queue is full                      queue: api_request_p1, queue_len: 1024, msg: the system has insufficient resources for load, service: bproc
Jun 10 10:04:28.260 ERRO Work queue is full                      queue: api_request_p1, queue_len: 1024, msg: the system has insufficient resources for load, service: bproc
Jun 10 10:04:28.260 ERRO Work queue is full                      queue: api_request_p1, queue_len: 1024, msg: the system has insufficient resources for load, service: bproc
Jun 10 10:04:28.260 WARN Error processing HTTP API request       method: GET, path: /eth/v1/beacon/blocks/1842608/attestations, status: 500 Internal Server Error, elapsed: 4.359789ms
Jun 10 10:04:28.260 WARN Error processing HTTP API request       method: GET, path: /eth/v1/beacon/blocks/1842608/attestations, status: 500 Internal Server Error, elapsed: 4.583824ms
Jun 10 10:04:28.260 WARN Error processing HTTP API request       method: GET, path: /eth/v1/beacon/blocks/1842608/attestations, status: 500 Internal Server Error, elapsed: 4.536422ms
Jun 10 10:04:28.278 ERRO Work queue is full                      queue: api_request_p1, queue_len: 1024, msg: the system has insufficient resources for load, service: bproc
Jun 10 10:04:28.278 WARN Error processing HTTP API request       method: GET, path: /eth/v1/beacon/blocks/1842608/attestations, status: 500 Internal Server Error, elapsed: 158.585µs
Jun 10 10:04:28.279 ERRO Work queue is full                      queue: api_request_p1, queue_len: 1024, msg: the system has insufficient resources for load, service: bproc
Jun 10 10:04:28.279 WARN Error processing HTTP API request       method: GET, path: /eth/v1/beacon/blocks/1842608/attestations, status: 500 Internal Server Error, elapsed: 57.001µs
```

We hope the community can investigate and resolve this unexpected issue.

**Comment by michaelsproul** [comment]
We believe this is already fixed by 5.2 but are investigating. Please upgrade to 5.2 and let us know if you see it again

**Comment by chong-he** [comment]
This could be related to the slashing event on Holesky. Another report on Lighthouse Discord: https://discord.com/channels/605577013327167508/605577013331361793/1250171631590969394

Mass slashing event occurring around the same time (if the log is in UTC, are you sure the log timestamp is UTC+8 as Lighthouse logs in UTC)?
https://holesky.beaconcha.in/slot/1842610

Anyway, as Michael mention, upgrading to 5.2 should help

**Comment by michaelsproul** [comment]
Closing this issue as high memory usage on v5.1.x is explained by poor state management in `verify_attester_slashing_for_gossip`, here:

https://github.com/sigp/lighthouse/blob/f1d88ba4b1256e084d3a2a547e0ce574b896246c/beacon_node/beacon_chain/src/beacon_chain.rs#L2491

With 5.2 and tree-states, cloning that state no longer consumes much memory.

**Comment by hzxGoForward** [comment]
> We believe this is already fixed by 5.2 but are investigating. Please upgrade to 5.2 and let us know if you see it again

Ok! We will upgrade to 5.2 and test it.

**Comment by hzxGoForward** [comment]
> This could be related to the slashing event on Holesky. Another report on Lighthouse Discord: https://discord.com/channels/605577013327167508/605577013331361793/1250171631590969394
> 
> Mass slashing event occurring around the same time (if the log is in UTC, are you sure the log timestamp is UTC+8 as Lighthouse logs in UTC)? https://holesky.beaconcha.in/slot/1842610
> 
> Anyway, as Michael mention, upgrading to 5.2 should help

Thank you for your invaluable advice; we will proceed with the upgrade to version 5.2.
